### PR TITLE
Opencascade 7 fixes

### LINKED ISSUE #2983: opencascade 7.0 compatibility
from https://github.com/Homebrew/homebrew-science/pull/4063

>  the opencascade 7.0.0 release is missing some headers dealii expects to be there, and similar such issues with some symbols renames, etc.

supposedly the same will happen to OCE...

### LINKED ISSUE #3915: Tests opencascade/arclength_boundary plane_intersection fail for opencascade 6.9.1 and later
The two tests
        7321 - opencascade/arclength_boundary_02.debug (Failed)
        7322 - opencascade/arclength_boundary_02.release (Failed)
        7357 - opencascade/plane_intersection_01.debug (Failed)
        7358 - opencascade/plane_intersection_01.release (Failed)

Fail for me with opencascade 6.9.1. It seems that some points get repeated in the output ??!
```
@@ -7,25 +7,25 @@
 4  1.00000 1.00000 0.00000
 5  0.500000 -0.207107 0.00000
 6  -0.207107 0.500000 0.00000
-7  -0.207107 0.500000 0.00000
+7  1.20711 0.500000 0.00000
 8  0.500000 1.20711 0.00000
-9  0.323223 0.500000 0.00000
+9  0.500000 0.500000 0.00000
 10  0.229402 -0.153281 0.00000
 11  0.770598 -0.153281 0.00000
 12  -0.153281 0.229402 0.00000
 13  -0.153281 0.770598 0.00000
-14  0.229402 -0.153281 0.00000
-15  0.229402 1.15328 0.00000
+14  1.15328 0.229402 0.00000
+15  1.15328 0.770598 0.00000
 16  0.229402 1.15328 0.00000
 17  0.770598 1.15328 0.00000
-18  0.411612 0.146447 0.00000
-19  0.411612 0.853553 0.00000
-20  0.0580583 0.500000 0.00000
-21  0.0580583 0.500000 0.00000
-22  0.145238 0.189433 0.00000
-23  0.385723 0.141597 0.00000
-24  0.145238 0.810567 0.00000
-25  0.385723 0.858403 0.00000
+18  0.500000 0.146447 0.00000
+19  0.500000 0.853553 0.00000
+20  0.146447 0.500000 0.00000
+21  0.853553 0.500000 0.00000
+22  0.189433 0.189433 0.00000
+23  0.810567 0.189433 0.00000
+24  0.189433 0.810567 0.00000
+25  0.810567 0.810567 0.00000
 $ENDNOD
 $ELM
 16
```

### PR DESCRIPTION
**Pull request by tamiko**
All changes tested with versions 6.9.1 and 7.1.0.

I get the following failing tests for 7.1.0 (see also #3915):

<del>2 - opencascade/arclength_boundary_02.debug (Failed)</del>
<del>6 - opencascade/circle_normal_projection.debug (Failed)</del>
<del>7 - opencascade/closest_point_to_shape.debug (Failed)</del>
<del>20 - opencascade/plane_intersection_01.debug (Failed)</del>

Closes #2983
Closes #3915

### CODE REVIEW

File: source/opencascade/boundary_lib.cc [hunk 1]
```diff
@@ -56,12 +59,12 @@ namespace OpenCASCADE
...
       if (shape.ShapeType() == TopAbs_WIRE)
-        return (Handle(BRepAdaptor_HCompCurve(new BRepAdaptor_HCompCurve(TopoDS::Wire(shape)))));
+        return Handle(BRepAdaptor_HCompCurve)(new BRepAdaptor_HCompCurve(TopoDS::Wire(shape)));
       else if (shape.ShapeType() == TopAbs_EDGE)
-        return (Handle(BRepAdaptor_HCurve(new BRepAdaptor_HCurve(TopoDS::Edge(shape)))));
+        return Handle(BRepAdaptor_HCurve)(new BRepAdaptor_HCurve(TopoDS::Edge(shape)));
 
       Assert(false, ExcInternalError());
-      return Handle(BRepAdaptor_HCurve(new BRepAdaptor_HCurve()));
+      return Handle(BRepAdaptor_HCurve)(new BRepAdaptor_HCurve());
```
Comment (tamiko) [inline source/opencascade/boundary_lib.cc]: @luca-heltai It is fascinating how the old version (with the wrong parenthesis) worked by accident :-D

File: doc/news/changes/minor/20170321Matthiasmaier [hunk 2]
```diff
@@ -0,0 +1,3 @@
+Fixed: Configugartion and compilation with opencascade 7
```
Comment (drwells) [inline doc/news/changes/minor/20170321Matthiasmaier]: typo (configuration)

**Comment by tamiko** [comment]
@davydden Jup, will do.

**Comment by tamiko** [comment]
@davydden I have tested oce-0.18, which is opencascade 6.9.1 with a community patchset. Fully compatible, no broken tests (well apart from the two that canonically fail).

Further, I have addressed the two additional failing tests.

**Comment by drwells** [comment]
I had to add `TKSTEP209` to the list of required libraries to get the tests to run with the current AUR [1] build of opencascade7 on Arch Linux. As long as I include that library I get the same set of test failures as you reported.

[1]The build instructions are available here https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=opencascade7 It is possible that we built things strangely, but nothing really sticks out  as unusual in the build script.

**Comment by tamiko** [comment]
@drwells TKSTEP209 seems to be present in 7* and 6.9*. I simply added it to the required link interface.

Does that fix the problem?

**Comment by drwells** [comment]
Yes: things work (up to the two known failing tests) perfectly for me with this library inclusion and version 7.1.0.

File: source/opencascade/boundary_lib.cc (hunk 1 above)
Comment (luca-heltai) [inline source/opencascade/boundary_lib.cc]: ah! good catch! :)

**Comment by tjhei** [comment]
So we are good to go?

**Comment by luca-heltai** [comment]
I'm on #3915 right now. If I manage to understand what's wrong with them, and fix them, we can close both. Give me 20 minutes. If not we can merge this and postpone those failing tests to after 8.5.

**Comment by tamiko** [comment]
@luca-heltai We can also give you an hour or two :-D

**Comment by luca-heltai** [comment]
So, with OCE-0.17 i get for arclength_boundary_02.debug the following mesh:

image

while with opencascade 7.1.0 I get 

image

**Comment by tamiko** [comment]
@luca-heltai So that is fixed now with the change of the handle type?

**Comment by luca-heltai** [comment]
No. That fixes the same two tests that you had already fixed on my compiler, which is not capable of convertiing handle(Geom_Circle) to handle(Geom_Curve).

**Comment by luca-heltai** [comment]
There might be differences in the way the new opencascade uses information from edge orientation and periodicity.

I'm not sure there is an easy fix. The two failing tests are producing similar results, i.e., it looks like the edge with periodicity is not traversed, and the bspline is going through the other side. 

I'd postpone this until we understand what is wrong with this. This does not affect deal.II users, unless they *construct* directly opencascade objects, like edges, curves, circles, etc., as I was doing in these tests to have quick shapes...

Lets postpone the fix to after 8.5 is released.

**Pull request merged**